feat: Web3 RPC Handler (npm package)

Keyrxng opened the pull request:
Resolves https://github.com/ubiquity/rpc-handler/issues/3

My package is hosted at ``yarn @keyrxng/web3-rpc-racer`` which I'm using in the video where I had left off the now closed pr.

- I've cleaned up the code
- Now using tsc to export types properly
- Cache refresh is configured when initializing
- Removed ts-template skeleton 
- Few more tests


> We should import the Chainlist repository as a submodule to stay in sync.

> The URLs should be tested upon initialization (i.e. page load) via Promise.race

> The class should rank sort based on latency.

> The class should drop failed endpoints.

> The class should have a wrapper method to make a network request using the optimal RPC endpoint, and to drop it/automatically rotate to the next best from the list if a network request fails.

> Make a standalone and polymorphic (frontend and backend compatible) npm module.

> Make sure to have unit tests in Jest.

All covered bar one

The wrapper method I haven't implemented as the spec has described it. The optimalRPC is outdated from call to call so ``getFastestRpcProvider()`` will retest either ``allNetworkUrls`` or just the successful cached rpcs, returning an updated provider which the invoking logic would handle making the call. This way the requested provider is always the quickest other than when calling ``getProvider()`` which will return the provider as-is.

the runtime rpcs are dropped as they fail for whatever reason so there is no need for rotation plus you always receive back the lowest latency provider.


##### Usage 

```ts
const rpcHandler = RPCHandler.getInstance(networkId, 5);
return rpcHandler.getProvider();
```

##### just shows the bad rpcs being excluded until refresh

https://github.com/ubiquity/web3-rpc-racer/assets/106303466/44566338-4dd3-45a8-9b55-54b925e66305

Keyrxng committed:
feat: web3 rpc handler npm package

Keyrxng committed:
feat: web3 rpc handler npm package

Keyrxng commented:
dont hold back @molecula451

0x4007 commented on .github/workflows/build.yml:
```diff
@@ -28,8 +28,8 @@ jobs:
         # env: # Set environment variables for the build
-          # SUPABASE_URL: "https://wfzpewmlyiozupulbuur.supabase.co"
-          # SUPABASE_ANON_KEY: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6IndmenBld21seWlvenVwdWxidXVyIiwicm9sZSI6ImFub24iLCJpYXQiOjE2OTU2NzQzMzksImV4cCI6MjAxMTI1MDMzOX0.SKIL3Q0NOBaMehH0ekFspwgcu3afp3Dl9EDzPqs1nKs"
+        # SUPABASE_URL: "https://wfzpewmlyiozupulbuur.supabase.co"
```
I guess it would be best not to publish these

0x4007 commented on build/esbuild-build.ts:
```diff
@@ -1,33 +1,65 @@
-];
+import chainlist from "../lib/chainlist/constants/extraRpcs";
+
+const typescriptEntries = ["rpc-handler.ts", "constants.ts", "handler.ts", "services/RPCService.ts", "services/StorageService.ts"];
```
I'm surprised it passed kebab case test?

Keyrxng replied:
- https://github.com/ubiquity/ts-template/tree/development/.github/workflows

ts-template doesn't have the kebab workflow in it by default, pay.ubq does tho. Just copy paste it over?

molecula451 replied:
> * https://github.com/ubiquity/ts-template/tree/development/.github/workflows
> 
> ts-template doesn't have the kebab workflow in it by default, pay.ubq does tho. Just copy paste it over?

bring it over https://github.com/ubiquity/web3-rpc-racer/actions/runs/8123062862/job/22205701380?pr=1

0x4007 commented on dist/cjs/constants.d.ts:
```diff
@@ -0,0 +1,24 @@
+  Anvil = 31337,
+}
+export declare enum Tokens {
+  DAI = "0x6b175474e89094c44da98b954eedeac495271d0f",
```
It would make more sense to associate tokens per network in case they aren't matching

Keyrxng replied:
Makes sense, should I add any more tokens in btw?

0x4007 replied:
Not for now but I suppose this should be passed in as constructor arguments instead so that the application can handle this?

0x4007 commented on src/services/StorageService.ts:
```diff
@@ -0,0 +1,28 @@
+export class StorageService {
+  static getLatencies(env: string): Record<string | number, number> {
+    if (env === "browser") {
+      return JSON.parse(localStorage.getItem("rpcLatencies") || "{}");
```
I don't think local storage modification should be within this module. 

Instead the results should be stored on the object. The end user should manually handle local storage if it's necessary I think.

Keyrxng replied:
So it's both, I felt that would be better. The invoking logic doesn't need to do anything besides initialize and the rest is taken care of, it saves having to repeat code across multiple UI's and for node envs the state is still saved onto the object

what would you suggest, that I remove this and have each UI handle it's own storage?

I'd need to expose the runtimeRPCSs and latencies, then the UI would pull those from the object after init and then save them into local. Anytime getFastestProvider is called the UI will pull from local, pass them into the handler then save the output to local again?

0x4007 replied:
You can consider decoupling the logic if you don't want each UI to handle it redundantly. Just have the localStorage logic in a separate method. 

However I would argue that it may not be necessary to use localStorage at all. If we do a Promise.race once upon page load  every time that's probably not bad and will accommodate for changing rpc conditions on a sporadic basis. 

Since all the RPC endpoints fire off in parallel we shouldn't have any bottle necks initializing like this.

Keyrxng replied:
Like a boolean passed into init enabling local storage handling automatically then methods that just set/get local storage?

The only problem I think is, with every page refresh the class is re-initialized and because the class state of latencies is <= 1 then it'll always race all RPCs. On init I pull the latencies and cache refresh count from localStorage and with passing the refresh arg in on init the UI is able to maintain a cache, so I think local is needed unless I'm not seeing something

We already initialize like you say essence just with the cached rpcs (unless refreshing the list), I can't think how we'd persist state across page visits without using local

0x4007 replied:
Yes I just think its not really necessary and adds complexity to making this polymorphic, or also being able to run in node.

Keyrxng committed:
chore: configurable, add kebab, remove dist

rndquu reviewed (changes requested):
@Keyrxng Pls add a proper readme file which describes:
- what the package does
- how to install it
- how to use it
- how to run tests

Keyrxng committed:
chore: readme

Keyrxng commented:
idk what the official name will be but I just went with rpc-handler

molecula451 commented:
build still fails @Keyrxng

Keyrxng committed:
chore: add workflow scripts

Keyrxng commented:
> build still fails @Keyrxng

Added the kebab scripts (forgot to when I added the workflow).

Deploying to cloudflare should be removed as there is nothing to deploy i.e no static folder which fails the run

and I think Knip is still not fully set up yet, is that right?

molecula451 commented on .github/workflows/scripts/kebab-case.sh:
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
```
the power of the generative pre-trained 🤣

0x4007 replied:
> the power of the generative pre-trained 🤣

I don't understand

molecula451 replied:
too much bash cryptic in this file, it only made sense that it was done by GPT, as a helpful, it's good tho!

0x4007 replied:
GPT is the future. Also I made this script lol

molecula451 commented:
> and I think Knip is still not fully set up yet, is that right?

yes you can skip/ignore knip

Keyrxng committed:
chore: remove build workflow

Keyrxng commented:
Is the run failing maybe because the workflow isn't on the branch it's being merged into? I have no idea why it would fail when it's a 1:1 copy of the stuff on the pay.ubq repo

molecula451 commented:
> Is the run failing maybe because the workflow isn't on the branch it's being merged into? I have no idea why it would fail when it's a 1:1 copy of the stuff on the pay.ubq repo

you mean the build.yml?

Keyrxng commented:
> > Is the run failing maybe because the workflow isn't on the branch it's being merged into? I have no idea why it would fail when it's a 1:1 copy of the stuff on the pay.ubq repo
> 
> you mean the build.yml?

The kebab case workflow, the debug run was a big help

molecula451 commented:
> The kebab case workflow, the debug run was a big help

yeah we are on a permissions issue: cc @gitcoindev 
![Screenshot from 2024-03-04 11-50-12](https://github.com/ubiquity/web3-rpc-racer/assets/41552663/40762440-72f3-46a3-9f64-2ab1847d9482)

0x4007 committed:
Delete .github/workflows/scripts/kebabalize.sh

This is destructive and renames files so lets not use it.

rndquu commented:
@Keyrxng 

Trying to run this code:
```
import { RPCHandler } from "./src";

async function main() {
    const handler = new RPCHandler(100);    
}

main();
```

Getting this error:
```
rndquu@Vladimirs-MBP web3-rpc-racer % npx tsx my.ts
/Users/rndquu/Public/projects/backend/web3-rpc-racer/src/constants.ts:78
  [networkIds.Mainnet]: ["https://rpc-pay.ubq.fi/v1/mainnet", ...(extraRpcs[networkIds.Mainnet] || [])],
                                                                  ^


ReferenceError: extraRpcs is not defined
    at extraRpcs (/Users/rndquu/Public/projects/backend/web3-rpc-racer/src/constants.ts:78:67)
    at Object.<anonymous> (/Users/rndquu/Public/projects/backend/web3-rpc-racer/src/constants.ts:85:27)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Object.S (/Users/rndquu/Public/projects/backend/web3-rpc-racer/node_modules/tsx/dist/cjs/index.cjs:1:1292)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at <anonymous> (/Users/rndquu/Public/projects/backend/web3-rpc-racer/src/index.ts:1:131)
    at Object.<anonymous> (/Users/rndquu/Public/projects/backend/web3-rpc-racer/src/index.ts:16:28)

Node.js v20.11.1
```

What am I doing wrong?

Keyrxng commented:
You are importing the unbuilt version, import from /dist/cjs it should work

rndquu commented:
@Keyrxng 

Trying to run this code:
```
import { RPCHandler } from "./src";

async function main() {
    const handler = new RPCHandler(100);    
}

main();

```

Getting this error:
```
rndquu@User-MBP web3-rpc-racer % npx tsx my.ts
/Users/rndquu/Public/projects/backend/web3-rpc-racer/src/constants.ts:78
  [networkIds.Mainnet]: ["https://rpc-pay.ubq.fi/v1/mainnet", ...(extraRpcs[networkIds.Mainnet] || [])],
                                                                  ^


ReferenceError: extraRpcs is not defined
    at extraRpcs (/Users/rndquu/Public/projects/backend/web3-rpc-racer/src/constants.ts:78:67)
    at Object.<anonymous> (/Users/rndquu/Public/projects/backend/web3-rpc-racer/src/constants.ts:85:27)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Object.S (/Users/rndquu/Public/projects/backend/web3-rpc-racer/node_modules/tsx/dist/cjs/index.cjs:1:1292)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at <anonymous> (/Users/rndquu/Public/projects/backend/web3-rpc-racer/src/index.ts:1:131)
    at Object.<anonymous> (/Users/rndquu/Public/projects/backend/web3-rpc-racer/src/index.ts:16:28)

Node.js v20.11.1

```

What am I doing wrong?

rndquu commented:
> You are importing the unbuilt version, import from /dist/cjs it should work

Thanks, I'll try

molecula451 commented:
> What am I doing wrong?

extraRpcs it's part of chainlist submodule make sure you git recursively clone or init the module

rndquu commented:
@Keyrxng 

I've installed git submodules via `git submodule update --init --recursive` and rebuilt the project.

Trying to get RPCs for BNB chain via:
```
import { RPCHandler } from "./dist/cjs/rpc-handler";

async function main() {
    const handler = new RPCHandler(56);    
}

main();
```

Getting this error:
```
TypeError: Cannot read properties of undefined (reading '0')
    at new r (/Users/rndquu/Public/projects/backend/web3-rpc-racer/dist/cjs/rpc-handler.js:19:6622)
    at main (/Users/rndquu/Public/projects/backend/web3-rpc-racer/my.ts:4:21)
    at <anonymous> (/Users/rndquu/Public/projects/backend/web3-rpc-racer/my.ts:7:1)
    at Object.<anonymous> (/Users/rndquu/Public/projects/backend/web3-rpc-racer/my.ts:7:6)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Object.S (/Users/rndquu/Public/projects/backend/web3-rpc-racer/node_modules/tsx/dist/cjs/index.cjs:1:1292)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at cjsLoader (node:internal/modules/esm/translators:356:17)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:305:7)
```

What am I doing wrong?

rndquu commented on README.md:
````diff
@@ -1,8 +1,84 @@
+- Re-test the cached RPCs
+
+```typescript
+const provider = handler.getFastestRpcProvider();
````
```suggestion
const provider = await handler.getFastestRpcProvider();
```

rndquu commented:
@Keyrxng 

This code returns the same provider all the time (`https://rpc-pay.ubq.fi/v1/mainnet`). Why is it happening? Is `https://rpc-pay.ubq.fi/v1/mainnet` so fast?

```
import { RPCHandler } from "./dist/cjs/rpc-handler";

async function main() {
    const handler = new RPCHandler(1);   
    const provider = handler.getProvider();
    console.log(provider);
}

main();
```

0x4007 commented:
> https://rpc-pay.ubq.fi/v1/mainnet

I just deleted this while adjusting the DNS settings yesterday. I presume that the code isn't handling the case when a provider goes down successfully.

Keyrxng committed:
chore: improved race, tests and readme, removed ubq rpcs

Keyrxng committed:
chore: script permission

Keyrxng commented:
> > https://rpc-pay.ubq.fi/v1/mainnet
> 
> I just deleted this while adjusting the DNS settings yesterday. I presume that the code isn't handling the case when a provider goes down successfully.

No it was handling this, I was testing mostly through the UI and relying on localStorage. Realized the tests weren't very robust if I were to exclude the browser testing, so had to update the promise racing and the tests

@rndquu everything is working now (never tested before with just a script (just the fact it was running in jest tests was proof enough it was node compat but that was a half-arsed approach) but I've covered all bases this time, my apologies)

Keyrxng commented:
> This code returns the same provider all the time (`https://rpc-pay.ubq.fi/v1/mainnet`). Why is it happening? Is `https://rpc-pay.ubq.fi/v1/mainnet` so fast?
> 
> ```
> import { RPCHandler } from "./dist/cjs/rpc-handler";
> 
> async function main() {
>     const handler = new RPCHandler(1);   
>     const provider = handler.getProvider();
>     console.log(provider);
> }
> 
> main();
> ```

So the init function was async but the constructor isn't, I was assigning the first provider in the constructor, so it would have been returning the first one every time in a node env because the init function was calling the async testRpcPerformance and never getting the response

init() just updates the config now, the user needs to test the RPCs by calling one of the two getFastestRPC or TestRpcPerformance. I have updated the readme to match the correct flow for both envs

<hr/>


https://github.com/ubiquity/web3-rpc-racer/assets/106303466/1451e93d-ae52-4409-8e34-1134a2bfbd26

gentlementlegen commented on src/constants.ts:
```diff
@@ -0,0 +1,69 @@
+  return networkName ?? "Unknown Network";
+}
+
+export const networkExplorers: Record<number, string> = {
```
Might make more sense to replace `number` by the key of network ids (applies for the other variables as well) to make sure every key is implemented

Keyrxng replied:
Is what I have now 'too typed'? 

Thought best to pull in all of the network IDs and names but had to restructure their type to match the networkNames structure

gentlementlegen commented on src/rpc-handler.ts:
```diff
@@ -0,0 +1,212 @@
+    } else if (networkId !== 31337 && this._provider?.connection.url.includes("localhost")) {
+      this._provider = await this.testRpcPerformance(networkId);
+    } else {
+      this._provider = await this.testRpcPerformance(networkId);
```
else if and else are identical

molecula451 replied:
it's idiomatic to put the last else, as is, tho

0x4007 commented on build/esbuild-build.ts:
```diff
@@ -43,7 +43,7 @@ esbuild
     tsconfig: "tsconfig.web.json",
     platform: "browser",
-    outdir: "dist/esm",
+    outdir: "dist/esm/src",
```
Why don't you output both in the same format? I guess cjs is the most common?

0x4007 replied:
> if I were to exclude the browser testing

By definition this code isn't polymorphic if you have to write separate tests for node and browser. You should make the code simpler to maintain by only writing one version that works in both environments.

Keyrxng replied:
The difficulty I was having which forced my hand was dealing with the extraRpcs declaration not being included in the build with its data tho this amongst over things fixed that

Keyrxng replied:
> Having distinct build outputs for ESM and CJS in a package is a common and recommended practice for maximizing compatibility and performance across different JavaScript environments. It requires careful configuration but ultimately makes your package more versatile and future-proof.

https://chat.openai.com/share/f1a72763-fe0d-4092-a1eb-e5166b63853c

0x4007 commented on src/rpc-handler.ts:
```diff
@@ -27,20 +27,66 @@ export class RPCHandler implements HandlerInterface {
+  public async getFastestRpcProvider(networkId: number): Promise<JsonRpcProvider> {
+    if (networkId === 31337) {
+      this._provider = new JsonRpcProvider(LOCAL_HOST, networkId);
+    } else if (networkId !== 31337 && this._provider?.connection.url.includes("localhost")) {
```
```suggestion
    } else if (this._provider?.connection.url.includes("localhost")) {
```

Keyrxng committed:
chore: type the constants remove bad else

Keyrxng commented:
The networkID !== 31337 and includes("localhost") was a fix for an edge case where on cycle refresh it would respond with a localhost:8545 while it should have been a real RPC but thats been fixed now so yeah no need for the elsif

0x4007 commented on README.md:
````diff
@@ -1,8 +1,85 @@
+## Installation
+
+```bash
+yarn add @ubiquity/rpc-handler
````
We should rename this repo to match.

molecula451 replied:
done renamed

0x4007 commented on jest.config.js:
jest.config.ts please

https://github.com/ubiquity/ts-template/pull/19/files

rndquu commented:
@Keyrxng Check this code:
```
import { HandlerConstructorConfig } from "./dist/cjs/src/handler";
import { RPCHandler } from "./dist/cjs/rpc-handler";

const config: HandlerConstructorConfig = {
    autoStorage: true,
    cacheRefreshCycles: 5,
  };

async function main() {
    const networkId = 100;
    const handler = new RPCHandler(networkId, config);
    const provider = await handler.getFastestRpcProvider(networkId);
    console.log(provider);
}

main();
```

Questions:
1. The code above returns `http://localhost:8545` as the best provider all the time, is it expected?
2. Why do we need to pass `networkId` both to `new RPCHandler(networkId, config)` and `handler.getFastestRpcProvider(networkId)`? Can't we set this config param only 1 time in a single place?

Keyrxng committed:
fix: type out path to src

Keyrxng commented:
I see that you are importing from ``dist/cjs/rpc-handler`` and not ``dist/src/cjs/rpc-handler``, I doubled check and I had the type.d out path wrong (was working out compile errors because in order for the tests to run they rely on the `/dist` folder but it's not there for a fresh pull, thought I covered it)

The returning of the localhost provider was partly the reason for that else if scenario but in my manual testing via the UI, the tests and manually running the script that has since been fixed.

I assume you are just trying to run the script with tsx or something?

> Why do we need to pass networkId both to new RPCHandler(networkId, config) and handler.getFastestRpcProvider(networkId)?

I suppose we don't need to, my original thinking was that it might be useful to be able to pass in any ID


```ts
import { HandlerConstructorConfig } from "./handler";
import { RPCHandler } from "../dist/cjs/src/rpc-handler";

const config: HandlerConstructorConfig = {
  autoStorage: true,
  cacheRefreshCycles: 5,
};

async function main() {
  const networkId = 100;
  const handler = new RPCHandler(networkId, config);
  return await handler.getFastestRpcProvider(networkId);
}

main().then(console.log).catch(console.error);
```

![image](https://github.com/ubiquity/rpc-handler/assets/106303466/ed2d21df-5912-44dd-ba2d-bfc2c938bcbd)



https://github.com/ubiquity/rpc-handler/assets/106303466/9f84e499-6369-4981-ae48-f9468b385463

Keyrxng committed:
chore: type the promises, add ts config and jest workflow

rndquu commented:
> I assume you are just trying to run the script with tsx or something?

Yes

Now I'm getting the `Module '"./dist/cjs/src/handler"' has no exported member 'HandlerConstructorConfig'` error

<img width="1512" alt="Screenshot 2024-03-06 at 16 11 50" src="https://github.com/ubiquity/rpc-handler/assets/119500907/89f14c77-0c92-417b-8ae1-fc9cbe45b02a">

How to reproduce:
1. Checkout to the latest commit from this PR
2. `yarn install`
3. `git submodule update --init --recursive`
4. `yarn build`
5. `npx tsx my.ts`

```ts
import { HandlerConstructorConfig } from "./dist/cjs/src/handler";
import { RPCHandler } from "./dist/cjs/src/rpc-handler";

const config: HandlerConstructorConfig = {
    autoStorage: true,
    cacheRefreshCycles: 5,
  };

async function main() {
    const networkId = 56;
    const handler = new RPCHandler(networkId, config);
    const provider = await handler.getFastestRpcProvider();
    console.log(provider);
}

main();
```

Keyrxng committed:
chore: final touch ups

Keyrxng commented:
Your screenshot made me realize two things:

- type.d out was not right
- I had added all the ids etc at the type level but no further

I've tested multiple times with fresh vsc instance, yarn, build, test and script and everything is good for me, it should be for you too

- removed axios and just using built-in fetch
- `config` is the only arg passed in now and the only mandatory prop is `networkId`
- removed unused stuff like network explorer, currencies, tokens etc as they serve no purpose in racing the RPCs but a good thing to be exported from the package if it's going to be used across many repos
- fixed the tests
- found that it's the `JsonRpcProvider` that is returning the `http:/\/localhost:8545";` so I've added the check back in

Keyrxng committed:
chore: export all from index

Keyrxng committed:
fix: readme

rndquu reviewed (changes requested):
1. If you build the project with `yarn build` then this line `import { HandlerConstructorConfig } from "./dist/cjs/src/handler";` returns the `Module '"./dist/cjs/src/handler"' has no exported member 'HandlerConstructorConfig'` error. But if you build with `yarn build:cjs` there is no such error. Pls make sure that `yarn build`, `yarn build:cjs` and `yarn build:esm` are working the same way.
2. Getting this error:
```
rndquu@Users-MacBook-Pro web3-rpc-racer % npx tsx my.ts
node:internal/modules/cjs/loader:1144
  const err = new Error(message);
              ^

Error: Cannot find module './dist/cjs/src/rpc-handler'
Require stack:
- /Users/rndquu/Public/projects/backend/web3-rpc-racer/my.ts
    at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at a._resolveFilename (/Users/rndquu/Public/projects/backend/web3-rpc-racer/node_modules/tsx/dist/cjs/index.cjs:1:1729)
    at Module._load (node:internal/modules/cjs/loader:985:27)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at <anonymous> (/Users/rndquu/Public/projects/backend/web3-rpc-racer/my.ts:2:28)
    at Object.<anonymous> (/Users/rndquu/Public/projects/backend/web3-rpc-racer/my.ts:16:6)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Object.S (/Users/rndquu/Public/projects/backend/web3-rpc-racer/node_modules/tsx/dist/cjs/index.cjs:1:1292)
    at Module.load (node:internal/modules/cjs/loader:1207:32) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/Users/rndquu/Public/projects/backend/web3-rpc-racer/my.ts' ]
}

Node.js v20.11.1
```
How to reproduce:
1. Checkout to the latest commit from this PR
2. `yarn install`
3. `git submodule update --init --recursive`
4. `yarn build:cjs`
5. `npx tsx my.ts`

```ts
import { HandlerConstructorConfig } from "./dist/cjs/src/handler";
import { RPCHandler } from "./dist/cjs/src/rpc-handler";

const config: HandlerConstructorConfig = {
    networkId: 100,
    autoStorage: false,
    cacheRefreshCycles: 5,
  };

async function main() {
    const handler = new RPCHandler(config);
    const provider = await handler.getFastestRpcProvider();
    console.log(provider);
}

main();
```

Keyrxng commented:
Your build dir should look like this after running ``yarn build``

![image](https://github.com/ubiquity/rpc-handler/assets/106303466/d6722d9c-bb82-4319-bdcb-e54b87443ff2)

``yarn build`` is the only yarn command that should be run as postbuild handles the types

Vid below is me doing a fresh ``yarn`` > ``yarn build`` > ``npx tsx ./tests/script-test.ts``

I do not understand why you are still getting this error @rndquu 

https://github.com/ubiquity/rpc-handler/assets/106303466/158e2e83-2873-4d01-ae8d-6c80d73de8b6

0x4007 commented:
Just make a GitHub action to prove its not env related?

rndquu commented:
@Keyrxng I haven't touched anything in the code since yesterday but today the build is performed without errors

Check this script:
```ts
import { HandlerConstructorConfig } from "../dist/cjs/src/handler";
import { RPCHandler } from "../dist/cjs/src/rpc-handler";

const config: HandlerConstructorConfig = {
  networkId: 314,
  autoStorage: false,
};

async function main() {
  const handler = new RPCHandler(config);
  const provider = await handler.getFastestRpcProvider();
  console.log(provider);
}

main();
```

It returns:
```
rndquu@User-MBP web3-rpc-racer % npx tsx tests/script-test.ts
[RPCService] Latencies object is empty
[RPCService] Failed to find fastest RPC
```

Although there are available RPCs for the Filecoin network:
<img width="1512" alt="Screenshot 2024-03-11 at 10 03 52" src="https://github.com/ubiquity/rpc-handler/assets/119500907/a869983d-1c56-4fa8-bcf8-43580c57410e">

Why is it happening?

Keyrxng commented:
Happy days we are  on the same page with it now tho good stuff

It's failing because I have the request timeout at 500ms, only three rpcs are returned for 314 that have `data-tracking-none`. Upping the timeout to 1500ms it succeeded, which raises a few questions doesn't it?

We want to keep the timeout low as in the sense that UBQ use it but it might be an idea for me to implement some sort of fallback where if none succeed then try again without the timeout? As we use 100 and 1 and they have lots of RPCS for both those chains the timeout should be kept in as I've seen calls take ^3s, so the timeout keeps things clean when there are lots of endpoints to test but in the case of there being very few endpoints or few good quality endpoints maybe it should be handled? Our use-case and the needs that UBQ have for it are covered anything else is more of a public service than anything I think is it not? :))

Let me know how to handle it and I'll push it today

```bash
$ npx tsx ./tests/script-test.ts
chainID 314:  filecoin
chain:  Filecoin
rpcss:  [
  'https://api.node.glif.io',
  'https://node.filutils.com/rpc/v1',
  'https://api.chain.love/rpc/v1'
]
Trace: r {
  _isProvider: true,
```

```yml
Provider available
Block number: 3728033
{
  'https://node.filutils.com/rpc/v1_314': 789.8341999999999,
  'https://api.node.glif.io_314': 1249.8516,
  'https://api.chain.love/rpc/v1_314': 1176.7801
}
```

rndquu commented:
> It's failing because I have the request timeout at 500ms
>
> that have `data-tracking-none`

You're right, ubiquity case is covered, all these settings (timeout, `data-tracking-none`) could be moved to config in a separate issue

rndquu reviewed (approved):
Works fine

Keyrxng commented:
> > It's failing because I have the request timeout at 500ms
> > that have `data-tracking-none`
> 
> You're right, ubiquity case is covered, all these settings (timeout, `data-tracking-none`) could be moved to config in a separate issue

I was going to suggest just that, add timeout etc into config and leave it to the user

gentlementlegen commented on package.json:
```diff
@@ -1,23 +1,32 @@
 {
-  "name": "ts-template",
+  "name": "@ubiquity/rpc-handler",
```
Should be `@ubiquitydao/rpc-handler`

molecula451 replied:
but the org it's named **ubiquity** here on github

0x4007 replied:
> but the org it's named **ubiquity** here on github

We just created https://www.npmjs.com/org/ubiquitydao and this will be deployed to npm there!

gentlementlegen replied:
Sadly @ubiquity already exists but no one seems to know who owns it

0x4007 commented on package.json:
```diff
@@ -1,23 +1,32 @@
 {
-  "name": "ts-template",
+  "name": "@ubiquity/rpc-handler",
```
```suggestion
  "name": "@ubiquitydao/rpc-handler",
```

0x4007 committed:
Update package.json

molecula451 commented on README.md:
````diff
@@ -1,8 +1,87 @@
+## Installation
+
+```bash
+yarn add @ubiquity/rpc-handler
````
remame here too accordingly

molecula451 replied:
if people do yarn add @ubiquity/ will get 404, do `"@ubiquitydao/rpc-handler"` considering the package it's gonna be hosted there, i rember i renamed just to match the repo name

molecula451 committed:
Update README.md

molecula451 commented:
Tested, working good, it looks like https://gnosis-pokt.nodies.app it's making the race at this time of writing, the README can be improved further with all the stuff that was tested here, as it it only available as a **NPM** installation on the current readme

![rpc](https://github.com/ubiquity/rpc-handler/assets/41552663/37db9f13-f540-40d6-a2d1-6986730eeeda)

molecula451 commented:
Code looks good and well structured, the package it's working well, there are maybe things to improve that will only become handful as the package it's use, mergin this, good stuff @Keyrxng
